### PR TITLE
Create search fragment

### LINKED ISSUE #7: Search support in navigation
**Is this a BUG REPORT or FEATURE REQUEST?**:
feature

**What happened**:
No support for search in navigation.

**What you expected to happen**:
Support for search in navigation.

**How to reproduce it (as minimally and precisely as possible)**:
Look for search support in navigation.

**Anything else we need to know?**:
Look into index.js + lunr.js

**Environment**:
- Syna Theme version:  v0.7.0
- Hugo version:  v0.30.2

### LINKED ISSUE #8: Create search fragment
**Is this a BUG REPORT or FEATURE REQUEST?**:
feature

**What happened**:
No search fragment.

**What you expected to happen**:
Simple search fragment.

**How to reproduce it (as minimally and precisely as possible)**:
Look for a search fragment.

**Anything else we need to know?**:
Look into index.js + lunr.js

**Environment**:
- Syna Theme version:  v0.7.0
- Hugo version:  v0.30.2

### PR DESCRIPTION
**Pull request by mpourismaiel**
**What this PR does / why we need it**:
Add search fragment. Suitable for usage in a separate page. Can be used inside navigation with a few style changes in the future.

**Which issue this PR fixes**:
fixes #7
fixes #8

**Special notes for your reviewer**:
Demo is available in [here](https://deploy-preview-327--syna-demo.netlify.com/search). Almost all of the code is here thanks to [eddiewebb](https://gist.github.com/eddiewebb/735feb48f50f0ddd65ae5606a1cb41ae).

The config is generated using `fuse.js`'s [website](http://fusejs.io/). It might seem weird so please feel free to change it. Fuse is a fuzzy search and in that, it seems a bit buggy. Not really sure if it's buggy or not.

Please do not merge this so we can have a bit more discussion on it (possibility of changing search library, using an online service and ...).

**Release note**:
```release-note
- search: Add search fragment
```

### CODE REVIEW

File: layouts/partials/fragments/nav.html [hunk 1]
```diff
@@ -20,6 +23,23 @@
         {{- .Params.asset.text -}}
       </a>
     {{- end }}
+    {{- if .Params.search }}
+      <div class="search-container mr-3">
```
Comment (stp-ip) [inline layouts/partials/fragments/nav.html]: Why not use "form-inline" from bootstrap? https://getbootstrap.com/docs/4.1/components/navbar/

File: layouts/partials/fragments/nav.html [hunk 2]
```diff
@@ -81,3 +101,30 @@
     Top
   {{- end -}}
 </div>
+{{- if .Params.search }}
+<!-- this template is sucked in by search.js and appended to the search-results div above. So editing here will adjust style -->
+<script id="search-result-template-{{ .Name }}" type="text/x-js-template">
+  <div id="summary-${key}">
+    <h5><a href="${link}">${title}</a></h5>
```
Comment (stp-ip) [inline layouts/partials/fragments/nav.html]: I would link the whole container and not just the title.
In more detail: Make the whole container for each search result clickable, but show only the title as actual link.

File: exampleSite/content/search/index.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
++++
```
Comment (stp-ip) [inline exampleSite/content/search/index.md]: If the example is meant to be used to showcase for the user later too, we should start with using `fragments/search` right away.

File: exampleSite/config.toml [hunk 4]
```diff
@@ -9,6 +9,9 @@ enableGitInfo = true
 version = "0.12.0"
 disableKinds = ["RSS", "taxonomy", "taxonomyTerm"]
 
+[outputs]
+  home = ["HTML", "RSS", "JSON"]
```
Comment (stp-ip) [inline exampleSite/config.toml]: Not sure RSS is necessary as we have disabled this `kind` above.

File: layouts/_default/index.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  {{- $.Scratch.Set "tags" nil -}}
+  {{- $.Scratch.Set "categories" nil -}}
+  {{- range first 1 (where ($.Scratch.Get "pages") "Params.fragment" "eq" "content") -}}
+    {{- $.Scratch.Set "plain" .Plain -}}
+    {{- $.Scratch.Set "tags" .Params.tags -}}
+    {{- $.Scratch.Set "categories" .Params.categories -}}
+  {{- end -}}
+  {{- $.Scratch.Add "index" (dict "title" .Title "tags" ($.Scratch.Get "tags") "categories" ($.Scratch.Get "categories") "contents" ($.Scratch.Get "plain") "permalink" .Permalink) -}}
+{{- end -}}
+{{- $.Scratch.Get "index" | jsonify -}}
```
Comment (stp-ip) [inline layouts/_default/index.json]: Interesting solution to print the actual index.

File: layouts/partials/fragments/search.html [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            {{- . | markdownify -}}
+          </h6>
+        </div>
+      </div>
+    {{- end }}
+    <div class="row">
+      <form action="{{ "search" | absURL }}" class="w-100">
+        <div class="col-md-6 col-12">
+          <div class="form-group row">
+            <label for="search-query" class="col-2 pl-0 pt-2">Search:</label>
```
Comment (stp-ip) [inline layouts/partials/fragments/search.html]: Wouldn't provide a label as we already have `title` and `subtitle`.

A few other design ideas:
* center by default
* possibility for left/right align
* use search icon similar to navbar
* match color/style (needs separate discussion)

File: layouts/partials/fragments/search.html [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            <input type="text" class="form-control col-10 col-md-8" id="search-query-{{ .Name }}" name="s" placeholder="Search for something...">
```
Comment (stp-ip) [inline layouts/partials/fragments/search.html]: Placeholder should use `i18n`.

File: layouts/partials/fragments/search.html [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+          </div>
+        </div>
+      </form>
+    </div>
+    <div class="row">
+      <div class="col px-0">
+        <h3>Matching pages</h3>
```
Comment (stp-ip) [inline layouts/partials/fragments/search.html]: `i18n`

File: layouts/partials/fragments/search.html [hunk 9]
```diff
@@ -0,0 +1,76 @@
+{{- $bg := .Params.background | default "light" -}}
+{{- $self := . -}}
+{{- .page_scratch.Add "js" ("syna-search.js" | relLangURL) }}
+
+{{ "<!-- Pricing -->" | safeHTML }}
```
Comment (stp-ip) [inline layouts/partials/fragments/search.html]: pricing -> search

File: layouts/partials/fragments/search.html [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+      <div class="col px-0">
+        <h3>Matching pages</h3>
+        <div id="search-results-{{ .Name }}" class="col px-0">
+        </div>
+      </div>
+    </div>
+  </div>
+</section>
+<!-- this template is sucked in by search.js and appended to the search-results div above. So editing here will adjust style -->
+<script id="search-result-template-{{ .Name }}" type="text/x-js-template">
```
Comment (stp-ip) [inline layouts/partials/fragments/search.html]: Maybe use a helper partial for the template so it's consistent for navbar and search fragment?

File: static-main/js/search.js [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+    $.ajax({ method: 'get', url: '/index.json' }).then(data => {
+      this.indexCache = data;
+      callback(data);
+    });
+  }
+
+  search(query) {
+    if (!query) {
+      return this.resultsContainer.html('<p>Please enter a word or phrase above</p>');
```
Comment (stp-ip) [inline static-main/js/search.js]: Probably hard to make that language agnostic later on. Aka we can't just use `i18n`.

File: static-main/js/search.js [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    this.getIndex(data => {
+      const pages = data;
+      const fuse = new Fuse(pages, this.fuseOptions);
+      const matches = fuse.search(query);
+      if (matches.length > 0) {
+        this.populateResults(matches, query);
+      } else {
+        this.resultsContainer.html('<p>No matches found</p>');
```
Comment (stp-ip) [inline static-main/js/search.js]: Same as above.

File: layouts/partials/head.html [hunk 13]
```diff
@@ -43,4 +43,8 @@
 
   {{- partial "custom/css.html" . }}
 
+  <script>
```
Comment (stp-ip) [inline layouts/partials/head.html]: What's the reasoning behind the code below?

**Comment by mpourismaiel** [comment]
- [ ] Move navigation to left if search is enabled, search would be aligned right
- [ ] Move search to collapse in mobile
- [ ] Extract design of search results section to a partial and share them between navbar and search fragment

File: layouts/partials/fragments/nav.html (hunk 1 above)
Comment (stp-ip) [reply layouts/partials/fragments/nav.html]: I still think it makes sense to use this. Might make upgrade to new bootstrap version less problematic and reduces styling we ourselves do.

File: layouts/_default/index.json (hunk 5 above)
Comment (mpourismaiel) [reply layouts/_default/index.json]: I cannot take credit for this. :D https://gist.github.com/eddiewebb/735feb48f50f0ddd65ae5606a1cb41ae

Comment (stp-ip) [reply layouts/_default/index.json]: Still interesting ^^

File: layouts/partials/fragments/nav.html (hunk 1 above)
Comment (mpourismaiel) [reply layouts/partials/fragments/nav.html]: I think there is no styling for this input right now. We use `input-group`. It's documented in https://getbootstrap.com/docs/4.1/components/forms/#inline-forms

**Comment by mpourismaiel** [comment]
As it might go amiss: https://github.com/okkur/syna/pull/327#discussion_r214436416

File: layouts/partials/fragments/nav.html [hunk 14]
```diff
@@ -4,7 +4,7 @@
   {{- .page_scratch.Add "js" ("syna-search.js" | relLangURL) -}}
 {{- end }}
 
-{{- "<!-- Navigation -->" | safeHTML -}}
+{{- "<!-- Search -->" | safeHTML -}}
```
Comment (stp-ip) [inline layouts/partials/fragments/nav.html]: Isn't it still Navigation, but with an optional search?

File: layouts/partials/fragments/search.html [hunk 15]
```diff
@@ -39,38 +39,32 @@ <h6>
...
-<script id="search-result-template-{{ .Name }}" type="text/x-js-template">
-  <div id="summary-${key}">
-    <h4><a href="${link}">${title}</a></h4>
-    <p>${snippet}</p>
-    ${ isset tags }<p>Tags: ${tags}</p>${ end }
-    ${ isset categories }<p>Categories: ${categories}</p>${ end }
-  </div>
-</script>
+{{ partial "helpers/search-result-template.html" (dict "name" .Name "message_class" "text-center mt-4") }}
+{{ partial "helpers/search-result-template.html" . }}
```
Comment (stp-ip) [inline layouts/partials/fragments/search.html]: I assume calling it twice populates the page and the other populates the overlay right?

Comment (stp-ip) [reply layouts/partials/fragments/search.html]: OK I'm missing something. What's the reason for the second partial call with the whole site data?

File: layouts/partials/fragments/nav.html (hunk 1 above)
Comment (stp-ip) [reply layouts/partials/fragments/nav.html]: Ok then happy to leave it as is.

**Comment by stp-ip** [comment]
Looks good. One clarification question and 2 todos still open: https://github.com/okkur/syna/pull/327#issuecomment-417459465

**Comment by stp-ip** [comment]
Also the alignment seems broken on the actual search page. https://deploy-preview-327--syna-demo.netlify.com/search/

**Comment by mpourismaiel** [comment]
I think everything is now in order. I updated the nav example with the search param but set it to false since navbar already has a search and two navbar with two searches will mess up :D

**Comment by stp-ip** [comment]
Awesome. We are almost there.

Found a small bug. Enter something in the navbar search -> hit enter or click search icon -> redirected to search page, but no search term is given.

**Comment by mpourismaiel** [comment]
Haha clicking on the submit button or pressing Enter triggered the close search functionality and form was submitted after that. Fixed the issue.

**Comment by stp-ip** [comment]
And we have search. That is so awesome. \o/

**Pull request merged**